### PR TITLE
ImageNet Timing Fixes

### PR DESCRIPTION
**Pull request by pomonam**
This PR aims to make the general PyTorch imagenet resnet workload more efficient. Specifically, this PR:
- Removes `fast_collate` and performs normalization with `torch.transform`.
- Removes sync batch norm. 
- Removes irrelevant functions.
- Updates PyTorch version (`1.12.1` --> `1.13.0`).

I confirmed that this reduces the timing for `load data` and `update_params`. Moreover, I ran full-scale experiments and confirmed that the changes still reach the target accuracy using target-setting hyperparameters. 

One minor thing is that increasing `num_workers` in the data loader seems to have a negative effect during the evaluation. I set `num_workers=0` for evaluation data loaders and this slightly slows down the evaluation. We may need to investigate this further in the future.

### CODE REVIEW

File: algorithmic_efficiency/workloads/imagenet_resnet/imagenet_v2.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def _decode_example(example: Dict[str, float]) -> Dict[str, float]:
                                                resize_size)
     return {'inputs': image, 'targets': example['label']}
 
-  ds = ds.map(_decode_example, num_parallel_calls=16)
```
Comment (znado) [inline algorithmic_efficiency/workloads/imagenet_resnet/imagenet_v2.py:40]: can we undo this change? or is it needed for some reason?

File: algorithmic_efficiency/workloads/imagenet_resnet/imagenet_pytorch/workload.py [hunk 2]
```diff
@@ -154,7 +155,6 @@ def init_model_fn(
     model.to(DEVICE)
     if N_GPUS > 1:
       if USE_PYTORCH_DDP:
-        model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
```
Comment (znado) [inline algorithmic_efficiency/workloads/imagenet_resnet/imagenet_pytorch/workload.py:157]: I'm a pytorch noob so I may be misunderstanding, but from my reading of the email thread, now that we delete the sync BN from here does that mean we _have_ to use DDP in order to get BN stats to sync at all? should we throw an error if `USE_PYTORCH_DDP == False` in the imagenet workloads?

**Review by znado — APPROVED** [review]

File: algorithmic_efficiency/workloads/imagenet_resnet/imagenet_v2.py (hunk 1 above)
Comment (pomonam) [reply algorithmic_efficiency/workloads/imagenet_resnet/imagenet_v2.py:40]: Sorry, I encountered a weird problem with `num_workers` for PyTorch and this was one of my attempts at debugging. I will undo this line.

File: algorithmic_efficiency/workloads/imagenet_resnet/imagenet_pytorch/workload.py (hunk 2 above)
Comment (pomonam) [reply algorithmic_efficiency/workloads/imagenet_resnet/imagenet_pytorch/workload.py:157]: In my understanding, `DataParallel ` updates the BN stats on the default device and we don't need to use synchronization (https://discuss.pytorch.org/t/understanding-behavior-of-torch-nn-dataparallel/91676).

File: algorithmic_efficiency/workloads/imagenet_resnet/imagenet_v2.py (hunk 1 above)
Comment (pomonam) [reply algorithmic_efficiency/workloads/imagenet_resnet/imagenet_v2.py]: Fixed!

**Comment by runame** [comment]
@pomonam I have noticed that the `DistributedEvalSampler` was removed. What is the reason for that? I have added it because when `drop_last=False` and `len(dataset) % N_GPUS != 0` `DistributedSampler` will add in additional examples, which leads to incorrect validation metrics (see [this discussion](https://github.com/pytorch/pytorch/issues/25162)).

**Comment by pomonam** [comment]
> @pomonam I have noticed that the `DistributedEvalSampler` was removed. What is the reason for that? I have added it because when `drop_last=False` and `len(dataset) % N_GPUS != 0` `DistributedSampler` will add in additional examples, which leads to incorrect validation metrics (see [this discussion](https://github.com/pytorch/pytorch/issues/25162)).

I thought that we added the condition to raise Exceptions in those cases, but it turns out that I was wrong and we were checking the other condition. I will add this function back in. Thank you for pointing it out!

**Comment by runame** [comment]
> I thought that we added the condition to raise Exceptions in those cases, but it turns out that I was wrong and we were checking the other condition. I will add this function back in. Thank you for pointing it out!

Ah I see, I have implemented padding for the workloads with tf input pipelines, so we don't have to check for this condition.

**Comment by pomonam** [comment]
Fixed! I incorporated all feedback.
